### PR TITLE
add dunder __dir__ for java methods

### PR DESCRIPTION
**Pull request by amsukdu**
For readable / usable propose, added `__dir__` to JavaMethod & JavaMultipleMethod. 
It is more human readable than JNI signatures.

Solves #419 issue
`readable_sig` is in .pxi but if it can be moved to .py anytime. I can make some test cases also.

### CODE REVIEW

**Comment by JPDSousa** [comment]
@amsukdu can we see the output of the `dir` methods? Maybe include it as a comment on the review. I'll try to test it out on my end anyway. Also, I'm not aware of the standards for this repo, but maybe some automatic testing is in order as well :)

**Comment by amsukdu** [comment]
@JPDSousa 
Yes, I can make some tests. I'll push it as soon as I wrote.

**Comment by tshirtman** [comment]
That looks cool! Sorry for very late feedback, i do think this is valuable, if you can, a few tests for the helper function that converts from jni to java notation, would be nice, i'll likely merge either way, and write them myself if you don't have time for it.

**Comment by amsukdu** [comment]
Hello, @tshirtman
I added more comments & fixed typos. I'm not 100% sure about what the helper function is referring to but I write the function `readable_sig` in cpython but it is basically the same as normal python code except `cdef`.

Your changes are welcome!

File: tests/test_dir.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+from jnius.reflect import autoclass
+
+class DirTest(unittest.TestCase):
+
+    def test_varargs_dir(self):
+        # >>> from jnius import autoclass
+        # >>> cls = autoclass('java.lang.System')
+        # >>> dir(cls.out.printf)
+        # [(['java/lang/String', 'java/lang/Object...'], 'java/io/PrintStream'),
+        # (['java/util/Locale', 'java/lang/String', 'java/lang/Object...'], 'java/io/PrintStream')]
```
Comment (tshirtman) [inline tests/test_dir.py]: why not test that this is true, in the test? maybe
```
assert (['java/lang/String', 'java/lang/Object...'], 'java/io/PrintStream') in dir(cls.out.printf)
```

etc?

File: tests/test_dir.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        #  (['double'], 'java/lang/String'),
+        #  (['float'], 'java/lang/String'),
+        #  (['int'], 'java/lang/String'),
+        #  (['java/lang/Object'], 'java/lang/String'),
+        #  (['long'], 'java/lang/String')]
+        cls = autoclass("java.lang.String")
+        assert isinstance(dir(cls.valueOf), list)
+        for f in dir(cls.charAt):
+            assert isinstance(f, tuple)
+
```
Comment (tshirtman) [inline tests/test_dir.py]:
```suggestion
```

File: tests/test_dir.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
```
Comment (tshirtman) [inline tests/test_dir.py]:
```suggestion
```

File: tests/test_dir.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
```
Comment (tshirtman) [inline tests/test_dir.py]:
```suggestion
```

File: tests/test_dir.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
```
Comment (tshirtman) [inline tests/test_dir.py]:
```suggestion
```

File: tests/test_dir.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
```
Comment (tshirtman) [inline tests/test_dir.py]:
```suggestion
```

**Comment by tshirtman** [comment]
I'm wondering now about the validity of replacing `__dir__` for java methods objects, wouldn't it be cleaner to have a specific attribute or method of the object, to call for instead of twisting the semantics of the dir() function for this object? https://docs.python.org/3/library/functions.html#dir 
> If the object has a method named `__dir__()`, this method will be called and must return the list of attributes. This allows objects that implement a custom `__getattr__()` or `__getattribute__()` function to customize the way dir() reports their attributes.

File: tests/test_dir.py (hunk 1 above)
Comment (amsukdu) [reply tests/test_dir.py]: I thought matching exact strings are miscellaneous but it seems more appropriate than just plain comments.

**Comment by amsukdu** [comment]
Yes, my first intention was piggybacking to `__dir__` method. The semantics doesn't perfectly match but I would say it is in the same ballpark with the original. The original `dir()` from `JavaMethod` was not helpful for me to understand the object. Many python users would poke around with `dir()` or `help()` so it could be a trade-off between strict policy and make another custom method, or a little bit of flexibility.

I'm totally okay with either way :)

**Comment by tshirtman** [comment]
dir indeed is useful to look into the object, but adding a `signature` method will also be visible in the result of dir() and i've no doubt that it'll get the attention of anyone looking for that information.

**Comment by amsukdu** [comment]
Agreed. Document this `signature` method to the `README` can be very helpful to the new comers.

**Comment by tshirtman** [comment]
Closing this one as i merged #467 thanks a lot!

**Pull request closed without merging**